gitlabform/configuration/projects_and_groups.py

```diff
@@ -18,61 +18,89 @@ def get_projects(self) -> list:
 
-        # this is simplistic, but for our config format should be enough for additive merge
-        # of project, group and common configs
-
-        # first merge common config with group configs
-        for key in group_config.keys() | common_config.keys():
+        project_config = self.get_project_config(group_and_project)
+        logging.debug("Project config: %s" % project_config)
 
-            if key in common_config and key not in group_config:
-                group_config[key] = common_config[key]
-            elif key in group_config and key not in common_config:
-                group_config[key] = group_config[key]
+        common_and_group_config = self.merge_configs(common_config, group_config)
+        logging.debug("Effective config common+group/subgroup: %s" % common_and_group_config)
+
+        effective_config = self.merge_configs(common_and_group_config, project_config)
+        logging.debug("Effective config common+group/subgroup+project: %s" % effective_config)
+
+        return effective_config
+
+    def get_effective_subgroup_config(self, subgroup):
+
+        #
+        # Goes through a subgroups hierarchy, from top to bottom
+        #
+        # "x/y/x" -> ["x", "x/y", "x/y/z"]
+        #
+        # ...and for each element after 1st generate effective config from previous effective one merged with current:
+        #
+        #              |     v       |
+        #              \---> a       |
+        #                    |       v
+        #                    \------>b = effective config to return
+        #
+        # ..where a = merged_config("x", "x/y") and b = merged_config(a, "x/y/z")
+        #
+
+        effective_config = {}
+        elements = subgroup.split('/')
+        last_element = None
+        for element in elements:
+            if last_element:
+                next_level_subgroup = last_element + '/' + element
+                next_level_subgroup_config = self.get_group_config(next_level_subgroup)
+                logging.debug("Config for '%s': %s"
+                              % (next_level_subgroup, next_level_subgroup_config))
+                effective_config = self.merge_configs(effective_config, next_level_subgroup_config)
+                logging.debug("Merged previous level config for '%s' with config for '%s': %s"
+                              % (last_element, next_level_subgroup, effective_config))
+                last_element = last_element + '/' + element
             else:
-                # overwrite common settings with group settings
-                group_config[key] = {**common_config[key], **group_config[key]}
+                effective_config = self.get_group_config(element)
+                logging.debug("First level config for '%s': %s" % (element, effective_config))
+                last_element = element
+
+        return effective_config
+
+    def merge_configs(self, more_general_config, more_specific_config) -> dict:
```

Comment:
maybe call them just `general_config` and `specific_config`?

Reply:
Just "general" and "specific" seem to be too absolute terms here.

For example if we have configs for group X, its subgroup Y and in it a project Z, then we first merge configs X and Y and then the result with config for Z. So X has "more general" config and Y has "more specific", but then the result of their merge becomes "more general" one.